lib/nm/index.ts

```diff
@@ -0,0 +1,121 @@
+import * as Promise from 'bluebird';
+import * as _ from 'lodash';
+import NetworkManagerTypes from './types';
+
+const dbus: any = require('dbus-native');
+
+const systemBus: any = dbus.systemBus();
+const service: string[] = [ 'org', 'freedesktop', 'NetworkManager' ];
+
+export interface WifiDevice {
+	devicePath: string;
+	type: number;
+}
+
+export interface AccessPoint {
+	apPath: string;
+	apSsid: string;
+}
+
+export class NetworkManager extends NetworkManagerTypes {
+	wifiDevice: WifiDevice;
+	accessPoints: AccessPoint[];
+	wifiActive: boolean;
+
+	constructor() {
+		super();
+
+		this.wifiActive = true;
+
+		this.getWifiDevice()
+		.then((device: WifiDevice) => {
+			const service = Promise.promisifyAll(systemBus.getService('org.freedesktop.NetworkManager'));
+			service.getInterfaceAsync(device.devicePath, 'org.freedesktop.NetworkManager.Device.Wireless')
+			.then((iface) => {
+				console.log('YOLO');
+				iface.on('AccessPointAdded', (...data) => {
+					console.log('YAY ACCESS POINT ADDED', ...data);
+				});
+				iface.on('AccessPointRemoved', (...data) => {
+					console.log('YAY ACCESS POINT REMOVED', ...data);
+				});
+				console.log('SBOLO');
+			})
+			.catch((err) => {
+				console.error('gnappo', err);
+			});
+		});
+	}
+
+	toggleWifi(value) {
+		return this.wifiActive = value;
+	}
+
+	getBus() {
+		return systemBus;
+	}
+
+	callMethod(path: string[]|string = []) {
+		return (iface: string[]|string = []) => ([method, signature]) => (params: any = []) => {
+			return new Promise((resolve, reject) => {
+				// {
+				//   path: `/org/freedesktop/NetworkManager/Devices/1`, // Object fs path
+				//   destination: service.join('.'), // ['org', 'freedesktop', 'NetworkManager'] // Object
+				//   iface: 'org.freedesktop.DBus.Properties', // Object interface
+				//   member: 'Get', // interface method
+				//   signature: 'ss', // interface method params types
+				//   body: ['org.freedesktop.NetworkManager.Device', 'DeviceType'], // [<interface to get property of>, <property name>] // params
+				// }
+				const command = {
+					path: _.isString(path) ? path : `/${service.concat(path).join('/')}`,
+					destination: service.join('.'),
+					interface: _.isString(iface) ? iface : service.concat(iface).join('.'),
+					member: method,
+					signature,
+					body: params,
+				};
+				console.log('command', command.path, command.interface, method, ...params);
+				systemBus.invoke(command,
+				function(err, ...results) {
+					if (err) {
+						return reject(err);
+					}
+					return resolve(...results);
+				});
```

Comment:
Weird indentation, also I guess this is the same as 
```typescript
return Bluebird.fromCallback((callback) => {
    systemBus.invoke(command, callback);
});
```

Reply:
What do you mean by weird indentation? For every opened `{` there's a newline.
Do you mean the curried parameters? If so, I could refactor them to look like:
```
callMethod = (path: string[]|string = []) =>
	(iface: string[]|string = []) =>
	([method, signature]) =>
	(params: any[] = []) =>
	Bluebird.fromCallback((callback) => {
		const command = {
			path: _.isString(path) ? path : `/${service.concat(path).join('/')}`,
			destination: service.join('.'),
			interface: _.isString(iface) ? iface : service.concat(iface).join('.'),
			member: method,
			signature,
			body: params,
		};
		systemBus.invoke(command, callback);
	});
```

Which is actually nicer to me, looks a bit more functional too. Lmk